plugins/cli/__init__.py

```diff
@@ -22,234 +23,548 @@
-            self.la()
-        elif cmd == 'lo':
-            self.lo()
-        elif cmd == 'll':
-            self.lo()
-        elif cmd == 'lt':
-            self.lt()
-        elif cmd == 'cl':
-            self.cl()
-        elif cmd.startswith('update ') or cmd.startswith('up '):
-            self.update(cmd.lstrip('update').strip())
-        elif cmd.startswith('dump'):
-            self.dump(cmd.lstrip('dump').strip(), '*' in cmd or ':' in cmd)
-        elif cmd.startswith('tr'):
-            self.tr(cmd.lstrip('tr').strip())
-        elif cmd.startswith('rl'):
-            self.rl(cmd.lstrip('rl').strip())
-        elif cmd.startswith('rr'):
-            self.rr(cmd.lstrip('rr').strip())
-        elif cmd == 'rt' or cmd == 'runtime':
-            self.rt()
-        elif cmd == 'help' or cmd == 'h':
-            self.usage()
-        elif cmd in ('quit', 'q', 'exit', 'x'):
+
+        if self.__wait_for_password:
+            # We are waiting for a password. So we take the entered command as password
+            if Utils.create_hash(cmd) == self.hashed_password.lower():
+                self.logger.debug("CLI: {0} Authorization succeeded".format(self.source))
+
+                # Inform the client that we will not echo what has been entered. He will do this from now on
+                self.send(bytearray([0xFF, 0xFC, 0x01]))  # IAC WILL ECHO
+
+                self.push("Enter 'help' for a list of available commands.\n")
+                self.push("> ")
+                self.__wait_for_password = False
+                return
+            else:
+                self.logger.debug("CLI: {0} Authorization failed".format(self.source))
+                self.push("\r\nAuthorization failed. Bye\r\n")
+                self.close()
+                return
+
+        if cmd in ('quit', 'q', 'exit', 'x'):
             self.push('bye\n')
             self.close()
             return
+        else:
+            if not self.commands.execute(self, cmd, self.source):
+                self.push("Unknown command.\nEnter 'help' for a list of available commands.\n")
+
         self.push("> ")
 
-    def cl(self):
-        self.sh.log.clean(self.sh.now())
 
-    def ls(self, path, match=True):
-        if not path:
+class CLI(lib.connection.Server, SmartPlugin):
+    ALLOW_MULTIINSTANCE = False
+    PLUGIN_VERSION = '1.1.0'
+
+    def __init__(self, smarthome, update='False', ip='127.0.0.1', port=2323, hashed_password=''):
+        """
+        Constructor
+        :param smarthome: smarthomeNG instance
+        :param update: Flag: Updates allowed
+        :param ip: IP to bind on
+        :param port: Port to bind on
+        :param hashed_password: Hashed password that is required to logon
+        """
+        self.logger = logging.getLogger(__name__)
+
+        if hashed_password is None or hashed_password == '':
+            self.logger.error("CLI: You must set a password to use this plugin. CLI Plugin will not be initialized!")
```

Comment:
I knew that this one would cause a discussion. ;-))
In my opinion, security should be mandatory, this is why I made it this way. I think it's hard but it's reasonable. However I see your point that after an update the CLI plugin will not start and the reason for this might be somewhere deep in the log files.
How about option 
c) CLI Plugin will (in addition to the error message) start, but not allow to login, instead show a message and close the connection.